### PR TITLE
Methodwebtest: Migrate Existing Tools, Create New Tools, Update Documentation

### PR DESCRIPTION
**Pull request by payoub125**
# Done

- Migrated tools from `webscan` [PR](https://github.com/Method-Security/webscan/pull/87)
- Create new Injection CLIs that will enable specific information to be sent from the compiler

### CLI Fingerprint

```
% method general multi -h
Perform injection tests in the multiple locations of a target

Usage:
  methodwebtest general multi [flags]

Flags:
      --event string           The event type to test: XSSALERT, SQLIBOOLEAN, SQLIESCAPE, SQLITIMEDELAY
  -h, --help                       help for multi
      --injectionlocation string   The injection location to test: HEADER, PATH, QUERY, BODY, FORM, MULTIPART
      --method string              The HTTP method to use for the request
      --variabledata string        Json string of variable names and base values to add to injects

Global Flags:
  -o, --output string        Output format (signal, json, yaml). Default value is signal (default "signal")
  -f, --output-file string   Path to output file. If blank, will output to STDOUT
  -q, --quiet                Suppress output
      --retries int          Number of attempts per credential pair
      --sleep int            Sleep time between requests (seconds)
      --targets strings      The URL of target
      --timeout int          Timeout per request (seconds) (default 30)
  -v, --verbose              Verbose output
```

### Modules

- XSS_ALERT
- SQLI_BOOLEAN
- SQLI_TIMEDELAY
- SQLI_ESCAPE
- COMMAND_TIMEDELAY
- COMMAND_ECHO

## Migrated Tools

- `methodwebtest general header misconfigured`
- `methodwebtest general header serveroverload`
- `methodwebtest general header useragent`
- `methodwebtest general path crlf`
- `methodwebtest general path traversal`

- `methodwebtest apache path traversal`
- `methodwebtest apache path modfile`

- `methodwebtest nginx header bufferoverflow`
- `methodwebtest nginx path traversal`
- `methodwebtest nginx query reverseproxy`

## Fern Example

```
{
    "content": {
        "targets": [
            {
                "target": "https://example.com/image",
                "requestCount": 1,
                "attempts": [
                    {
                        "request": {
                            "baseUrl": "https://example.com",
                            "path": "/image",
                            "method": "GET",
                            "queryParams": {
                                "filename": "/etc/passwd"
                            },
                            "eventType": [
                                {
                                    "type": "PathEvent",
                                    "value": "TRAVERSAL"
                                }
                            ],
                            "statusCode": 404,
                            "responseBody": "BODY",
                            "responseHeaders": {
                                "Accept-Ranges": "bytes",
                                "Age": "378490",
                                "Cache-Control": "max-age=604800",
                                "Content-Type": "text/html",
                                "Date": "Wed, 18 Dec 2024 14:22:56 GMT",
                                "Etag": "\"1088432560+gzip\"",
                                "Expires": "Wed, 25 Dec 2024 14:22:56 GMT",
                                "Last-Modified": "Sat, 12 Oct 2024 03:26:51 GMT",
                                "Server": "ECAcc (dcd/7D24)",
                                "Vary": "Accept-Encoding",
                                "X-Cache": "404-HIT"
                            }
                        },
                        "finding": false,
                        "timeSent": "2024-12-18T09:22:56-05:00",
                        "timeReceived": "2024-12-18T09:22:56-05:00"
                    }
                ],
                "startTimestamp": "2024-12-18T09:22:56-05:00",
                "endTimestamp": "2024-12-18T09:22:56-05:00"
            }
        ],
        "config": {
            "targets": [
                "https://example.com/image"
            ],
            "paths": [
                "/etc/passwd"
            ],
            "queryParam": "filename",
            "responseCodes": "200-299",
            "ignoreBaseContent": true,
            "timeout": 30,
            "retries": 0,
            "sleep": 0,
            "successfulOnly": false,
            "type": "PathTraversalEngineConfig"
        }
    },
    "started_at": "0001-01-01T00:00:00Z",
    "completed_at": "2024-12-18T09:22:56.687829-05:00",
    "status": 0
}

```

### General

```
% method general -h
Perform general injection tests against a target

Usage:
  methodwebtest general [command]

Available Commands:
  header      Perform injection tests in the headers of a target
  multi       Perform injection tests in the multiple locations of a target
  path        Perform injection tests in the path of a target

Flags:
  -h, --help              help for general
      --retries int       Number of attempts per credential pair
      --sleep int         Sleep time between requests (seconds)
      --targets strings   The URL of target
      --timeout int       Timeout per request (seconds) (default 30)

Global Flags:
  -o, --output string        Output format (signal, json, yaml). Default value is signal (default "signal")
  -f, --output-file string   Path to output file. If blank, will output to STDOUT
  -q, --quiet                Suppress output
  -v, --verbose              Verbose output

Use "methodwebtest general [command] --help" for more information about a command.
```

### Apache

```
% method apache -h 
Perform apache specific injection tests against a target

Usage:
  methodwebtest apache [command]

Available Commands:
  path        Perform path injection tests against a target

Flags:
  -h, --help              help for apache
      --retries int       Number of attempts per credential pair
      --sleep int         Sleep time between requests (seconds)
      --targets strings   The URL of target
      --timeout int       Timeout per request (seconds) (default 30)

Global Flags:
  -o, --output string        Output format (signal, json, yaml). Default value is signal (default "signal")
  -f, --output-file string   Path to output file. If blank, will output to STDOUT
  -q, --quiet                Suppress output
  -v, --verbose              Verbose output

Use "methodwebtest apache [command] --help" for more information about a command.
```

### Nginx

```
% method general -h
Perform general injection tests against a target

Usage:
  methodwebtest general [command]

Available Commands:
  header      Perform injection tests in the headers of a target
  multi       Perform injection tests in the multiple locations of a target
  path        Perform injection tests in the path of a target

Flags:
  -h, --help              help for general
      --retries int       Number of attempts per credential pair
      --sleep int         Sleep time between requests (seconds)
      --targets strings   The URL of target
      --timeout int       Timeout per request (seconds) (default 30)

Global Flags:
  -o, --output string        Output format (signal, json, yaml). Default value is signal (default "signal")
  -f, --output-file string   Path to output file. If blank, will output to STDOUT
  -q, --quiet                Suppress output
  -v, --verbose              Verbose output

Use "methodwebtest general [command] --help" for more information about a command.
```

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -26,11 +26,9 @@ type MethodWebTest struct {
 	OutputSignal signal.Signal
 	RootCmd      *cobra.Command
 	VersionCmd   *cobra.Command
-	FuzzCmd      *cobra.Command
-	ProbeCmd     *cobra.Command
-	SpiderCmd    *cobra.Command
-	VulnCmd      *cobra.Command
-	AppCmd       *cobra.Command
+	GeneralCmd   *cobra.Command
```
Comment (sean-hacker) [inline cmd/root.go:29]: Same comment as the other PR. We don't actually need each command on the struct, just the root. We then just attach the various commands to the root command.

File: internal/general/multi/sqli/boolean.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		Retries:           config.Retries,
+		Sleep:             config.Sleep,
+	}
+
+	report := utils.RunMultiInjectionsEngine(ctx, &injectionConfig)
+	checkForSqliBooleanBased(report)
+	return report
+}
+
+func checkForSqliBooleanBased(report *methodwebtest.Report) {
```
Comment (sean-hacker) [inline internal/general/multi/sqli/boolean.go:40]: One thought for future improvements. This should also be an interface that allows us to iteratively add more complex logic for determining if an injection worked. Like you can imagine running through a list of implementations that would determine if an injection worked, perhaps even leveraging an edge LLM.

**Review by sean-hacker — APPROVED** [review]
@payoub125 , there's obviously a lot here. My main overarching suggestion on future improvements (don't block this, this PR is already large) is to have the detection logic become an interface and part of the engine such that we can have multiple detection mechanisms vote on whether there was in fact a successful detection.

File: cmd/root.go (hunk 1 above)
Comment (payoub125) [reply cmd/root.go:29]: ack, ill make a clean up ticket

File: internal/general/multi/sqli/boolean.go (hunk 2 above)
Comment (payoub125) [reply internal/general/multi/sqli/boolean.go:40]: I like this idea a lot

**Pull request merged**